flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -2086,6 +2104,36 @@ IntrinsicLibrary::genProduct(mlir::Type resultType,
                       "unexpected result for Product", args);
 }
 
+// RANDOM_INIT
+void IntrinsicLibrary::genRandomInit(llvm::ArrayRef<fir::ExtendedValue> args) {
+  assert(args.size() == 2);
+  Fortran::lower::genRandomInit(builder, loc, *args[0].getUnboxed(),
+                                *args[1].getUnboxed());
+}
+
+// RANDOM_NUMBER
+void IntrinsicLibrary::genRandomNumber(
+    llvm::ArrayRef<fir::ExtendedValue> args) {
+  assert(args.size() == 1);
+  auto harvest = builder.createBox(loc, args[0]);
```

Comment:
I am realizing we should add a `asBox` option to lower argument that are meant to be passed via descriptors, especially when they are intent(out).

The pattern `asAddr` + `createBox()` may end-up creating a descriptor of a temp instead of a descriptor for the variable itself.

If you try something like:
```
  real :: x(4) = [0., 0., 0., 0.]
  call random_number(x(1:4:2))
  print *, x
end
```
I am pretty sure this will print 0., 0., 0., 0. instead of random numbers in element 1 and 3.
This is because asAddr is calling `gen(expr)` that makes contiguous temps instead of descriptors for slices.

I do not know if we want this behavior to remain the default in the long term. But right now it would makes sense to me to have a `asBox` option to `Fortran::lower::LowerIntrinsicArgAs` that ends-up calling something like `genBoxArg` in ConvertExpr.cpp, and the `builder.createBox` would be useless here. That would avoid the temp.

It may be worth to start using it in other places that can take non contiguous array argument to avoid temps, though it is only necessary for intent(out) array arguments.

If what I described here is clear enough so that you can do it, go for it, otherwise, tell me and I will add the `asBox` option.

Reply:
Don't we create the box if the interface is explicit though? (I thought we'd fixed that, but it may still be TODO.)

Reply:
> Don't we create the box if the interface is explicit though? (I thought we'd fixed that, but it may still be TODO.)

In user function calls, yes, that's actually what `genBoxArg` was added for, but in intrinsic calls, the runtime interface is not inferred from the Fortran interface since each intrinsic may have a custom runtime signature. Instead, how an argument is lowered to is driven by the table in intrinsic lowering, and right now there is no box option, so it's probably time to add one.

Reply:
You are right Jean - output with an array section is wrong.  I've added a `LowerIntrinsicArgAs::Box` variant as at least a step in the suggested direction.  The array section test now generates correct output.